contrib/scalar/scalar.c

```diff
@@ -19,9 +19,27 @@ static int is_unattended(void) {
 	return git_env_bool("Scalar_UNATTENDED", 0);
 }
 
+static void strbuf_parentdir(struct strbuf *buf)
+{
+	int len = buf->len;
+	while (len > 0 && !is_dir_sep(buf->buf[--len]))
+		; /* keep looking for parent directory */
```

Comment:
This might be easier to read if the `--len` was pulled into the loop body instead of mutating the value in the condition. It will compile equivalently.

Reply:
Is there a clearer/concise way to do this? Because it's a pre-decrement, I'd either need to use a do-while loop or something like:

```
while (len > 0) {
    len--;
    if (is_dir_sep(buf->buf[len])) break;
}
```

which also has the added confusion of accessing `buf[len]`, but `len` isn't really representing the "length" at that point.

Reply:
Ah, the issue is to limit it to `len > 1` instead of `len > 0`.

```
while (len > 1 && is_dir_sep(buf->buf[len - 1]))
	len--;
```

Reply:
I think this still causes an "off-by-one" issue - for example, if you start with string `/abc/123`, the loop will continue until `len == 5` (because `buf[4] == '/'`). When the buffer length is reset at the end of the function, it will now include the separator at the end (since `len == 5`, so `strbuf_setlen(buf, len)` will set the buffer to `/abc/`). 

We can't blindly subtract 1 from the final length either, since the exit from the loop could be caused by the length check OR the directory separator check. Requiring that `len > 1` in the loop will break some existing behavior, since it wouldn't be able to handle `/` (which is needed in some of the iterative uses of the function).

Would it help if I added a more descriptive comment explaining what the loop was doing? e.g. "shorten the path string until it either the lowest subdirectory is removed (including path separator) or the string is empty"

Reply:
Sounds good. You clearly have thought more deeply about this than I have!

Reply:
Updated!

Reply:
I _think_ the canonical way in Git is to use `find_last_dir_sep()`, see e.g. https://github.com/git/git/blob/v2.32.0/exec-cmd.c#L232-L234

Reply:
Thanks, that worked perfectly!